pd/src/state/reader.rs

```diff
@@ -476,6 +477,108 @@ impl Reader {
         })
     }
 
+    /// Retrieve a stream of quarantined notes and their commitments, paired with the validator
+    /// identity key with which they are associated.
+    ///
+    /// If `maximum_unbonding_height` is `Some`, only notes whose unbonding height is less than or
+    /// equal to that height will be returned.
+    ///
+    /// If `validators` is `Some`, only notes which were associated with an undelegation from some
+    /// validator in that set will be returned. (This is more efficient than filtering after
+    /// receiving he stream, because the database is performing the filtration.)
+    pub fn quarantined_notes<'a>(
+        &self,
+        maximum_unbonding_height: Option<u64>,
+        validators: Option<impl IntoIterator<Item = impl Borrow<&'a IdentityKey>>>,
+    ) -> impl Stream<Item = Result<(IdentityKey, note::Commitment, NoteData)>> + Send + Unpin + '_
```

Comment:
This function is called during slashing and unbonding of notes; it's generalized so that it is applicable in both situations.